primitives/election-providers/src/onchain.rs

```diff
@@ -62,9 +69,10 @@ impl<T: Config> ElectionProvider<T::AccountId, T::BlockNumber> for OnChainSequen
 	type DataProvider = T::DataProvider;
 
 	fn elect() -> Result<Supports<T::AccountId>, Self::Error> {
-		let voters = Self::DataProvider::voters();
-		let targets = Self::DataProvider::targets();
-		let desired_targets = Self::DataProvider::desired_targets() as usize;
+		let (voters, _) = Self::DataProvider::voters(None).map_err(Error::DataProvider)?;
+		let (targets, _) = Self::DataProvider::targets(None).map_err(Error::DataProvider)?;
```

Comment:
sidenote: it feels a bit frustrating to have to query the voter again here instead of using the already cached on in multiphase, but I understand it because we want to reuse traits.

I think it is ok as the on chain should almost never be triggered.

At some point pallet-multi-phase could implement DataProvider itself. and instead of using staking data provider we could use multi-phase data provider for onchain election.

Reply:
Yeah, you totally have my logic reverse-engineered there. 

I think in the long term, the fallback should itself be another `ElectionProvider<DataProvider = MultiPhaseDataProvider>`, and then not only we don't query again, but we will also have a pretty nice nested system.

Reply:
I have this written in the `future plans` section of the module doc.